plugins/entando-plugin-jpdigital-exchange/src/main/java/org/entando/entando/aps/system/services/digitalexchange/install/ComponentInstaller.java

```diff
@@ -0,0 +1,197 @@
+
+@org.springframework.stereotype.Component
+public class ComponentInstaller implements ServletContextAware {
+
+    private static final Logger logger = LoggerFactory.getLogger(ComponentInstaller.class);
+
+    protected static final String COMPONENTS_DIR = "de_components";
+    protected static final boolean PROTECTED_RESOURCE = true;
+
+    private final DigitalExchangesClient client;
+    private final IStorageManager storageManager;
+    private final DatabaseManager databaseManager;
+    private final InitializerManager initializerManager;
+
+    private ServletContext servletContext;
+
+    @Autowired
+    public ComponentInstaller(DigitalExchangesClient client, IStorageManager storageManager,
+            DatabaseManager databaseManager, InitializerManager initializerManager) {
+        this.client = client;
+        this.storageManager = storageManager;
+        this.databaseManager = databaseManager;
+        this.initializerManager = initializerManager;
+    }
+
+    public void install(ComponentInstallationJob job, Consumer<ComponentInstallationJob> updater) {
+
+        double steps = 3;
+        int currentStep = 0;
+
+        downloadComponent(job);
+        job.setProgress(++currentStep / steps);
+        updater.accept(job);
+
+        installComponent(job);
+        job.setProgress(++currentStep / steps);
+        updater.accept(job);
+
+        reloadSystem();
+        job.setProgress(1);
+        job.setStatus(InstallationStatus.COMPLETED);
+        updater.accept(job);
+    }
+
+    private void downloadComponent(ComponentInstallationJob job) {
+
+        Path tempZipPath = null;
+
+        try {
+
+            tempZipPath = Files.createTempFile(job.getComponent(), "");
+
+            DigitalExchangeBaseCall call = new DigitalExchangeBaseCall(
+                    HttpMethod.GET, "digitalExchange", "components", job.getComponent(), "packages");
+
+            try (InputStream in = client.getStreamResponse(job.getDigitalExchange(), call)) {
+                Files.copy(in, tempZipPath, StandardCopyOption.REPLACE_EXISTING);
+            }
+
+            createComponentsDir();
+            extractZip(tempZipPath);
+
+        } catch (ApsSystemException | IOException | UncheckedIOException ex) {
+            logger.error("Error while downloading component", ex);
+            throw new InstallationException("Unable to save component", ex);
+        } finally {
+            if (tempZipPath != null) {
+                tempZipPath.toFile().delete();
+            }
+        }
+    }
+
+    private void createComponentsDir() throws ApsSystemException {
+        if (!storageManager.exists(COMPONENTS_DIR, PROTECTED_RESOURCE)) {
+            storageManager.createDirectory(COMPONENTS_DIR, PROTECTED_RESOURCE);
+        }
+    }
+
+    private void extractZip(Path tempZipPath) throws IOException {
+
+        ZipFile zip = new ZipFile(tempZipPath.toFile());
+
+        zip.stream().forEach(entry -> {
+
+            try {
```

Comment:
If you need to add a try block to close the zip file, this might get even more nested. Maybe it's better to extract this inner try block to another method to reduce complexity.